client/api/src/backend.rs

```diff
@@ -502,6 +514,14 @@ pub trait Backend<Block: BlockT>: AuxStore + Send + Sync {
 	/// Returns a handle to offchain storage.
 	fn offchain_storage(&self) -> Option<Self::OffchainStorage>;
 
+	/// Pin the block to keep bodies, justification and state available after pruning.
+	/// Number of pins are reference counted. Users need to make sure to perform
+	/// one call to `unpin_block` per call to `pin_block`.
+	fn pin_block(&self, hash: &Block::Hash, number: u64) -> sp_blockchain::Result<()>;
```

Comment:
Is number needed here?

Reply:
Yes, we need it because in the implementation we pin in state-db, which demands number https://github.com/paritytech/substrate/pull/13157/files/b22055306d04d614f8d1be12712b62d294be90b2#diff-0e523389a21c9bd9fc1da47816eb73a5ad21637337bb938abecedfe7f936d824R2602

Reply:
What I had in mind was that we could read this number internally, backend should know it (hash is should be enough to identify block).

Reply:
Ah got it. We could do that, but I am not sure it's worth it. At the call site, we already have the header and the number for free.

If we fetch it internally, in the worst case (not cached), we read the header from DB.

Reply:
Question if this optimization justifies the complexity of API. Also thinking in terms of reducing _block number_ in API (aka https://github.com/paritytech/substrate/issues/11292).

Reply:
Okay fair point, I will get rid of it :+1: